diff -Nurp -Nurp chromeos-installer-0.0.1.orig/installer/chromeos-postinst chromeos-installer-0.0.1/installer/chromeos-postinst
--- chromeos-installer-0.0.1.orig/installer/chromeos-postinst	2017-11-05 09:52:32.529238226 +0800
+++ chromeos-installer-0.0.1/installer/chromeos-postinst	2017-11-05 11:00:14.708764400 +0800
@@ -33,3 +33,10 @@ main() {
   chroot . /usr/bin/cros_installer postinst "/" "$@"
 }
 main "$@"
+
+# INSTALL_ROOT is the a temp dir that th new root fs mounts under
+# $@ is the device name of the new rootfs
+# We don't put below line in above main() to execute it in chroot because
+# our post install script needs to copy some files from current ROOT to the
+# new ROOT.
+${INSTALL_ROOT}/usr/sbin/flintos-post-install.sh ${INSTALL_ROOT}
