diff --git a/boot_control_chromeos.cc b/boot_control_chromeos.cc
index e9ad698..425f1c3 100644
--- a/boot_control_chromeos.cc
+++ b/boot_control_chromeos.cc
@@ -100,7 +100,8 @@ bool BootControlChromeOS::Init() {
   if (IsRemovableDevice(boot_disk_name_)) {
     LOG(INFO)
         << "Booted from a removable device, pretending we have only one slot.";
-    num_slots_ = 1;
+    //num_slots_ = 1;
+    num_slots_ = 2;
   } else {
     // TODO(deymo): Look at the actual number of slots reported in the GPT.
     num_slots_ = 2;
